main/src/library/pkger/Parser.flix

```diff
@@ -0,0 +1,108 @@
+    pub def pure(x: a): Parser[a] = input -> Some((input, x))
+
+    pub def ap(l: Parser[a -> b], r: Parser[a]): Parser[b] = input -> Option.flatMap(tup -> match tup { case (inputx, f) => Option.flatMap(tupx -> match tupx { case (inputxx, x) => Some((inputxx, f(x))) }, runParser(r, inputx)) }, runParser(l, input))
+
+    pub def liftA2(f: a -> b -> c, x: Parser[a], y: Parser[b]): Parser[c] = ap(fmap(f, x), y)
+
+    // Monad
+    pub def return(x: a): Parser[a] = pure(x)
+
+    // TODO: `flatMap` instead?
+    pub def bind(v: Parser[a], f: a -> Parser[b]): Parser[b] = input -> {
+        Option.flatMap(tup -> match tup { case (inputx, x) => runParser(f(x), inputx) }, runParser(v, input))
+    }
+
+    // Alternative
+    pub def empty(): Parser[a] = _ -> None
+
+    pub def choose(l: Parser[a], r: Parser[a]): Parser[a] = input -> Option.choose(runParser(l, input), runParser(r, input))
+
+    def cons(x: a, xs: List[a]): List[a] = x :: xs
+
+    def flipCons(xs: List[a], x: a): List[a] = x :: xs
+
+    pub def some(v: Parser[a]): Parser[List[a]] = cons <&> v <*> many(v)
+
+    pub def many(v: Parser[a]): Parser[List[a]] = some(v) <|> Parser.pure(Nil)
+
+    def listToString(xs: List[Char]): String = List.foldLeft((s, x) -> String.concat(s, Char.toString(x)), "", xs)
+
+    // Parser
+    pub def char(x: Char): Parser[char] = input -> match String.toList(input) {
+        case Cons(y, ys) => if (y == x) { Some((listToString(ys), x)) } else { None }
+        case Nil => None
+    }
+
+    def sequenceA(xs: List[Parser[a]]): Parser[List[a]] = List.reverse <&> List.foldLeft(liftA2(flipCons), pure(Nil), xs)
+
+    pub def string(s: String): Parser[String] = listToString <&> List.map(char,  String.toList(s)).sequenceA()
+
+    // Made easier if 1) `Parsec` was used 2) `String` was `List[Char]`
+    pub def span(p: Char -> Bool): Parser[String] = input -> {
+        let (l, r) = List.span(p, String.toList(input));
+        Some(listToString(r), listToString(l))
+    }
+
+    // TODO: Weird name.
+    pub def failOn(v: Parser[a], p: a -> Bool): Parser[a] = mapOption(v, x -> if (p(x)) { None } else { Some(x) })
+}
+
+// This is a large amount of namespace pollution. Right now this file is only loaded manually, so it's alright.
+pub def <&>(f: a -> b, parser: Parser[a]): Parser[b] = Parser.fmap(f, parser)
+
+pub def <&(x: a, parser: Parser[b]): Parser[a] = (_ -> x) <&> parser
+
+pub def &>(parser: Parser[a], x: b): Parser[b] = x <& parser
+
+pub def <*>(l: Parser[a -> b], r: Parser[a]): Parser[b] = Parser.ap(l, r)
+
+pub def <*(l: Parser[a], r: Parser[b]): Parser[a] = ((x, _) -> x) <&> l <*> r
+
+pub def *>(l: Parser[a], r: Parser[b]): Parser[b] = ((x -> x) <& l) <*> r
+
+pub def <|>(l: Parser[a], r: Parser[a]): Parser[a] = Parser.choose(l, r)
+
+pub def >>=(v: Parser[a], f: a -> Parser[b]): Parser[b] = Parser.bind(v, f)
+
+pub enum JsonValue {
+    case JsonNull
+    case JsonBool(Bool)
+    case JsonNumber(Int)
+}
+
+namespace JsonParser {
```

Comment:
I am concerned about the choice of JSON for two reasons:

1. It seems like a poor design choice overall.
2. It seems that it would require a significant effort to implement a correct JSON parser.

As for (1) there are plenty of online resources that discuss the idiosyncrasies of JSON, including (a) its lack of comments, (b) its number support, (c) its support for null/undefined, and (d) other peculiarities.

Reply:
I agree. I should've made it more explicit, but I'm simply using this to drive the implementation of a usable ParserCombinator. I didn't want to write a combinator, which I'm not too well versed in, *and* figure out a new format at the same time. This helps me drive some useful abstractions while only dealing with one unknown instead of two.

Agree that JSON is the wrong format to use. Is there a format you think would be better? I was thinking something like TOML, but maybe you have a better idea?

If I had more time I would've wanted to implement a [`Serde`-like](https://serde.rs/) framework that serializes to and from a common ADT.

Reply:
Not sure; we can talk about the format. I should point out that someone else is already working on a parser combinator library, but it will probably not be ready in time for you to use.